### PR TITLE
Create default account automatically

### PR DESCRIPTION
**Pull request by ChristophWurst**
fixes #166 

TODO:
- [x] save login password to session if config is available
- [x] create a default account if it does not exist
- [x] update the account password if it changed
- [ ] add unit tests
- [ ] find someone who tests this with LDAP
- [ ] document how to configure the default account

### CODE REVIEW

File: lib/service/defaultaccount/defaultaccountmanager.php [hunk 1]
```diff
@@ -0,0 +1,169 @@
...
+			return;
+		}
+
+		$accounts = $this->mapper->findByEmail($config->buildEmail($user), $user->getUID());
+		if (empty($accounts)) {
+			$this->createDefaultAccount($user, $config);
+		} else if (count($accounts) === 1) {
+			$this->updateDefaultAccount($accounts[0], $user, $config);
+		} else {
+			$this->logger->debug('unexpected state: more than one account with the default email found for user ' . $user->getUID());
```
Comment (ChristophWurst) [inline lib/service/defaultaccount/defaultaccountmanager.php]: if anybody knows a better solution for handling this unexpected state, let me know

**Comment by ChristophWurst** [comment]
@owncloud/mail @tahaalibra review please

**Comment by jancborchardt** [comment]
cc @blizzz for LDAP testing

**Comment by jancborchardt** [comment]
Yo yo @owncloud/mail folks come test this as well! :)

**Comment by ChristophWurst** [comment]
@LukasReschke sample config:
```
'app.mail.accounts.default' => [
            'email' => '%USERID%',
            'imapHost' => 'imap.nextcloud.com',
            'imapPort' => 123,
            'imapSslMode' => 'ssl',
            'smtpHost' = 'smtp.nextcloud.com';
            'smtpPort' = 321,
            'smtpSslMode' = 'ssl',
]
```

**Comment by jancborchardt** [comment]
So @lukasreschke @christophwurst do we merge this then? :)

**Comment by pierreozoux** [comment]
Managed to make it work \o/, I used this configuration:

```
  'app.mail.accounts.default' => [
    'email' => '%EMAIL%',
    'imapHost' => 'mail.indie.host',
    'imapPort' => 993,
    'imapUser' => '%EMAIL%',
    'imapSslMode' => 'ssl',
    'smtpHost' => 'mail.indie.host',
    'smtpPort' => 587,
    'smtpUser' => '%EMAIL%',
    'smtpSslMode' => 'tls',
  ],
```

**Comment by jancborchardt** [comment]
@pierreozoux great stuff! So it’s a +1 from you?

@LukasReschke @MorrisJobke @ChristophWurst can we merge then? ;)

**Comment by ChristophWurst** [comment]
> @LukasReschke @MorrisJobke @ChristophWurst can we merge then? ;)

Let me try to implement this properly from the beginning and create the account dynamically. As noticed by @LukasReschke we should try to circumvent storing the login password to the database.

**Comment by pierreozoux** [comment]
Do you think I can use your PR in prod? Or should I wait your new PR?

**Comment by ChristophWurst** [comment]
> Do you think I can use your PR in prod? Or should I wait your new PR?

I wouldn't use it as it may raise security issues as the password is stored in the DB and faulty accounts are not removed (in case something breaks)

**Pull request closed without merging**